### PR TITLE
Jasmine/Jest v20+: this syntax codemod

### LINKED ISSUE #57: Codemod this keyword usage
Newer versions of Jest doesn't support using `this` inside hooks (e.g. `beforeEach`). Instead variables defined outside the scope can achieve the same.

Although this project is mainly moving from other test runners and assertion libraries to Jest, we might be able to help people here...

Background https://github.com/facebook/jest/issues/3628#issuecomment-303339343

FYI @thymikee

### PR DESCRIPTION
**Pull request by avaly**
Closes #57

Context: https://github.com/facebook/jest/issues/3553

### CODE REVIEW

**Comment by skovhus** [comment]
@avaly thank you so much for contributing. 👍🏻

One thing I would like to figure out is how this new transformer fits with some of the other transformations. Example: when using Mocha, I guess it would make sense to run your transformer as part of the migration. So it this new transformer safe enough to run as part of the other codemods?

Let me know what you think. I'll review the code soon. : )

**Comment by avaly** [comment]
@skovhus This codemod assumes the source code already matches Jest's syntax (`describe`/`before`/`beforeEach`/`it`). So I would say it's safe to run as part of a migration as long as it's run as the last step in the process, after all the other codemods have converted the ava/mocha/tape syntax to Jest.

File: src/transformers/jasmine-this.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
+/**
+ * Codemod for transforming Jamine `this` context into Jest v20+ compatible syntax.
+ */
+
+import detectQuoteStyle from '../utils/quote-style';
+
+const testFunctionNames = ['after', 'afterEach', 'before', 'beforeEach', 'it'];
```
Comment (avaly) [inline src/transformers/jasmine-this.js]: Support also `test`

**Comment by skovhus** [comment]
Thanks for this! Looks good to me. : )

I think it might make sense to run this after the Mocha transformer, similar to how we run chai-should as part of should.js here https://github.com/skovhus/jest-codemods/blob/master/src/transformers/should.js#L89

File: src/cli/index.js [hunk 2]
```diff
@@ -77,6 +79,10 @@ inquirer
                     value: TRANSFORMER_CHAI_SHOULD,
                 },
                 {
+                    name: 'Jasmine: this Syntax',
```
Comment (skovhus) [inline src/cli/index.js]: I'm wondering a bit if we can come up with a more descriptive name... Should we call it "Jasmine: this usage"? Hm.

File: src/transformers/jasmine-this.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
+/**
+ * Codemod for transforming Jamine `this` context into Jest v20+ compatible syntax.
```
Comment (skovhus) [inline src/transformers/jasmine-this.js]: `Jasmine`

File: src/transformers/jasmine-this.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            const originalName = path.value.property.name;
+            const newName = originalName + 'Context';
+            replacedIdentifiersMap[originalName] = newName;
+            return j.identifier(newName);
+        };
+
+        const addLetDeclarations = () => {
+            Object.keys(replacedIdentifiersMap)
+                // Reverse the keys because we're adding them one by one to the front of the body array
+                .sort(reverseComparator)
```
Comment (skovhus) [inline src/transformers/jasmine-this.js]: I would just do `.sort().reverse()`

File: src/transformers/jasmine-this.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    const updateDescribes = () => {
+        return root
+            .find(j.CallExpression, {
+                callee: {
+                    type: j.Identifier.name,
+                    name: 'describe',
+                },
+            })
+            .filter(path => getValidThisExpressions(path.value).size() > 0)
+            .forEach(mutateDescribe)
```
Comment (skovhus) [inline src/transformers/jasmine-this.js]: Might make it a bit more readable by inlining `mutateDescribe` here...

**Comment by skovhus** [comment]
Can I get you to rebase master and use the new `finale` instead of quoting detection? See https://github.com/skovhus/jest-codemods/pull/39/files#diff-80cc69d0b8fb32fd19621a13564a36eeL218

**Comment by skovhus** [comment]
Be aware that Mocha has some magic this functions (see https://github.com/skovhus/jest-codemods/issues/56)

**Comment by avaly** [comment]
@skovhus I've address your comments, rebased, and added the special Mocha functions to an ignore list.

**Comment by probablyup** [comment]
Timeout actually has an equivalent in Jest/Jasmine. You can do this:

```js
it('does something', () => {
  // some assertion
}, 500);
```

Which sets a 500ms timeout. I don't think it's available for describe blocks though, so if you wanted to add support for this it'd need to be detected on the suite and then added to each `it`.

**Comment by AlanFoster** [comment]
Thanks for providing an awesome codeshift for this migration :+1:

For visibility I ran this locally, and ran into some issues.

This was my before code:

```js
const asyncCall = function () {
  return Promise.resolve({ foo: 10 });
};

describe('async example', function () {
  beforeEach(async function () {
    this.response = await asyncCall();
  });

  it('makes the call with token', async function () {
    expect(this.response).toEqual({ foo: 10 });
  });
});
```

And this was my after code:

```js
const asyncCall = function () {
  return Promise.resolve({ foo: 10 });
};

describe('async example', () => {
  let responseContext;
  beforeEach(() => {
    responseContext = await asyncCall();
  });

  it('makes the call with token', () => {
    expect(responseContext).toEqual({ foo: 10 });
  });
});
```

The after code no longer compiles:

```
  ● Test suite failed to run

     await is a reserved word (8:22)
         6 |   let responseContext;
         7 |   beforeEach(() => {
      >  8 |     responseContext = await asyncCall();
           |                       ^
         9 |   });
        10 |
        11 |   it('makes the call with token', () => {
```

i.e. The change from `async function` to using a fat arrow syntax means the `await` keyword is no longer valid.

It's possible to fix this by adding the `async` keyword:

```
  beforeEach(async () => {
    responseContext = await asyncCall();
  });
```

This may not be a problem with this particular code shift, but I thought i'd flag it up for visibility :+1:

**Comment by avaly** [comment]
@AlanFoster Thanks for reporting the issue. Indeed I overlooked async functions. I'll patch the codemod when I get a chance to treat async functions as well.

**Comment by AlanFoster** [comment]
@avaly No problem! Thanks for the code shift, it saved us a bunch of time :+1:

Out of interest - what are your thoughts on this approach:

Extract `this` variables to reference a `context` variable at the top, which can be cleared before every test.

So for one of your examples, we introduce a `context` variable at the top - which is accessible in the `beforeEach` and tests.

Current approach:

```js
describe('foo', () => {
    let barContext;
    let fooContext;
    beforeEach(() => {
        fooContext = { id: 'FOO' };
        barContext = { id: 'BAR', child: fooContext };
    });

    it('should have proper IDs', () => {
        expect(fooContext.id).to.equal('FOO');
        expect(barContext.id).to.equal('BAR');
        expect(barContext.child.id).to.equal('FOO');
    });
});
```

Shared context approach:

```js
describe('foo', () => {
    let context;
    beforeEach(() => {
      context = {};
    });    
   
    beforeEach(() => {
        context.foo = { id: 'FOO' };
        context.bar = { id: 'BAR', child: fooContext };
    });

    it('should have proper IDs', () => {
        expect(context.foo.id).to.equal('FOO');
        expect(context.bar.id).to.equal('BAR');
        expect(context.bar.child.id).to.equal('FOO');
    });
});
```

It stops the scenario of leaking test state between tests without realizing it, or tests passing by mistake.

**Comment by skovhus** [comment]
Thanks for the feedback @AlanFoster. And thanks @avaly for having a look at fixing it. 🤘🏻

**Comment by avaly** [comment]
@AlanFoster I've added the async function support. Feel free to give it another try.

I like your suggestion with the context object. I will give that a try later this week. Does anyone else have any better names for `context`?

**Comment by avaly** [comment]
I've updated the codemod using @AlanFoster's suggestion in https://github.com/skovhus/jest-codemods/pull/61#issuecomment-317451060 of grouping all context data in a shared object.

**Comment by AlanFoster** [comment]
@avaly The latest code shift works great! I was able to transform the entire codebase without any issues, and all tests pass :)

One thing that is catching my eye is the rewrite of `function () {` to ArrowFunctionExpressions within our codebase:

```diff
-    it('sets initial state', function() {
-      expect(this.state).toEqual({});
+    it('sets initial state', () => {
+      expect(testContext.state).toEqual({});
     });
```

This might be a jest best practice though? I haven't tried out any of the other jest-codemods, but I wasn't able to see any linting rules to enforce using fat arrows within description blocks within [eslint-plugin-jest](https://www.npmjs.com/package/eslint-plugin-jest) for instance

Let me know your thoughts if this code shift should preserve the original function type or always align with ArrowFunctionExpressions :+1:

**Comment by skovhus** [comment]
@AlanFoster thanks for testing it. 

I think using arrow functions makes sense for most people and is supported from node@4...

Back from vacation now, so let us get this released! @avaly thank you so much for your work.

**Comment by avaly** [comment]
@AlanFoster The main reason I chose to also replace normal functions with arrow functions is that since we no longer use `this` for context, `this` will point to some upper scope global inside arrow functions, thus making any `this` usage incorrect.

**Pull request merged**